swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt

```diff
@@ -0,0 +1,165 @@
+import com.netflix.spinnaker.swabbie.repository.ResourceTrackingRepository
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Component
+import java.time.Clock
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+
+/**
+ * A scheduled notification sender
+ */
+@Component
+@ConditionalOnExpression("\${swabbie.enabled:true}")
+class NotificationSender(
+  private val lockingService: LockingService,
+  private val resourceTrackingRepository: ResourceTrackingRepository,
+  private val notifier: Notifier,
+  private val clock: Clock,
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val notificationQueue: NotificationQueue,
+  private val registry: Registry,
+  private val workConfigurations: List<WorkConfiguration>
+) : DiscoveryAware() {
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val notificationsId = registry.createId("swabbie.notifications")
+
+  private val lockOptions = LockManager.LockOptions()
+    .withLockName(lockingService.ownerName)
+    .withMaximumLockDuration(lockingService.swabbieMaxLockDuration)
+
+  /**
+   * Periodically sends notifications to resolved resource owners
+   * The frequency of this scheduled action is cron based and configurable
+   * Defaults to every day at 9am.
+   */
+  @Scheduled(cron = "\${swabbie.notification.cron.schedule:0 0 9 1/1 * ?}")
+  fun sendNotifications() {
+    if (!isUp()) {
+      return
+    }
+
+    lockingService.acquireLock(lockOptions) {
+      val resources = resourceTrackingRepository.getMarkedResources()
+        .filter {
+          it.notificationInfo == null
+        }
+
+      groupRelatedResourcesThenNotify(resources)
+    }
+  }
+
+  /**
+   * @return Marked resources to notify on
+   */
+  private fun groupRelatedResourcesThenNotify(markedResources: List<MarkedResource>) {
+    val now = Instant.now(clock)
+    val tasks = mutableListOf<NotificationTask>()
+    do {
+      val notificationTask: NotificationTask? = notificationQueue.pop()
+      if (notificationTask == null) {
+        log.debug("Nothing to notify. Skipping...")
+        return
+      }
+
+      tasks.add(notificationTask)
+    } while (!notificationQueue.isEmpty())
+
+    tasks.forEach { task ->
+      val byResourceOwner = markedResources
+        .filter {
+          it.resourceType == task.resourceType
+        }.groupBy {
+          it.resourceOwner
+        }
+
+      byResourceOwner.forEach { (owner, list) ->
+        val resourcesAndConfigurations: List<Pair<MarkedResource, WorkConfiguration>> = list.toResourceAndConfigurationPairs()
+        val result = notifier.notify(
+          Notifier.Envelope(
+            recipient = owner,
+            resources = resourcesAndConfigurations
```

Comment:
What's the value in passing the whole work configuration here? Does the notification system need everything? If not maybe we could pass a lot less information.

Reply:
Ahh I didn't realize `Envelope` was only going to the `EchoNotifier`. To me envelope has the specific connotation of being used in a request/response body to wrap things. There's probably a clearer name for this.

Reply:
It doesn't have the same connotation to me. I am not married to that name so what's your suggestion?

Reply:
oof hit me back with the "what's your suggestion" 😆 good one! I'll think about it and try to offer one on my next pass of PR review.